File: leetcode/704.binary-search/ts/binarysearch.test.ts
import { expect, test } from "vitest";
import search from "./binarysearch.js";

interface testCase {
    nums: number[],
    target: number,
    want: number,
}

test("binarysearch", () => {
    const testcases: testCase[] = [
        { nums: [-1, 0, 3, 5, 9, 12], target: 9, want: 4 },
        { nums: [-1, 0, 3, 5, 9, 12], target: 2, want: -1 },
        { nums: [], target: 0, want: -1 },
        { nums: [1], target: 1, want: 0 },
        { nums: [2, 3], target: 3, want: 1 },
    ]

    for (const tc of testcases) {
        expect(search(tc.nums, tc.target)).toEqual(tc.want)
    }
})

File: leetcode/704.binary-search/cpp/binarysearch_test.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

#include "binarysearch.cpp"

struct testCase
{
    std::vector<int> nums;
    int target;
    int want;
};

TEST_CASE("BinarySearch")
{
    std::vector<testCase>
        tcs{
            {std::vector<int>{-1, 0, 3, 5, 9, 12}, 9, 4},
            {std::vector<int>{-1, 0, 3, 5, 9, 12}, 2, -1},
            {std::vector<int>{}, 0, -1},
            {std::vector<int>{1}, 1, 0},
            {std::vector<int>{2, 3}, 3, 1},
        };

    Solution s;

    for (auto tc : tcs)
    {
        CHECK(s.search(tc.nums, tc.target) == tc.want);
    }
}